### DISCUSSION TITLE
Server action has wrong typing for form action

### DISCUSSION THREAD

**Discussion by flotischoti** [vercel/next.js#56581] [Help] [answered]
### Summary

Hello,
I'm using a server action in a form in a server component . 
```
async function deleteUser(data: FormData) {
    'use server'
    ... do something ...
}

return <form action={deleteUser}> .... </form>
```
However, typescript indicates a type error on the form's `action` attribute: Type '(data: FormData) => Promise<void>' is not assignable to type 'string'.

`action` is defined as `action?: string | undefined` on interface `FormHTMLAttributes`. Since server functions need to be async and will therefore return a promise, I'm not sure how to resolve this. 

Any ideas?

### Additional information

```js
Next version: 13.5.4
React version: 18.2
```

**Comment by raferdev** [comment]
The form action receive an url (string) to send the data, i think to your purpose its better use onsubmit that's handle a function. Or create a route that receive the form data that with basically does the same thing.

**Reply by flotischoti** [reply]
`onSubmit` does not seem to be working on server components and is rather used together with `pages`. I'm using `app router` only and following exactly what the docs suggest: [server-only-forms](https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations#server-only-forms).
I prefer using server actions to api routes because in my experience they are faster and `revalidatePath()` method actually works here.

**Reply by raferdev** [reply]
You are right, i dont use server forms yet, testing here in my typescript its different: 

![exemple-form-action](https://github.com/vercel/next.js/assets/98192761/0610e2cc-b136-4547-8552-7cdb6cd98d40)

The types are working well, only if is something in the new next version. In other case you can declare react html element types to override it.

React version: 18.2
Next version: 13.4

**Reply by flotischoti** [reply]
Strange, I changed Next version to 13.4.x but still get the same problem.

How can I override the react html element? I tried the following but overriding doesn't seem to work while only adding new attributes has an effect:

```
// /index.d.ts
declare module 'react' {
  export interface FormHTMLAttributes {
    action?: string | ((data: FormData) => Promise<void>) | undefined
    test: string
  }
}
```
![image](https://github.com/vercel/next.js/assets/64696958/9768f6b5-92a1-485a-ab47-04bd28a22348)

**Reply by flotischoti** [reply, marked as answer]
Sorry, I'm dumb.

Just noticed that the react types were outdated. I didn't put a ^ in the semantic versioning in package.json.

**Reply by uptownhr** [reply]
@flotischoti what version did it upgrade to? Fresh install from create next and seeing this. 

```
"dependencies": {
    "@prisma/client": "^5.11.0",
    "next": "14.1.3",
    "prisma": "^5.11.0",
    "react": "^18",
    "react-dom": "^18"
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "autoprefixer": "^10.0.1",
    "postcss": "^8",
    "tailwindcss": "^3.3.0",
    "typescript": "^5"
  }
```
All I've done is copied example from https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#forms
<img width="831" alt="image" src="https://github.com/vercel/next.js/assets/1021110/32c0207d-4cbb-47b1-86dc-af4a80e6bd7d">

**Reply by flotischoti** [reply]
@uptownhr sorry for late reply. I haven't had access to my laptop for a couple of month.

`git list` shows:
├── @types/node@20.9.0
├── @types/react-dom@18.2.15
├── @types/react@18.2.37

with `package.json` content:
```
"@types/node": "^20",
"@types/react": "^18",
"@types/react-dom": "^18",
```
